src/sampler.jl

```diff
@@ -0,0 +1,120 @@
+                reltol_dd=0.05, abstol_dd=1e-2)
+    
+    # Linearly spaced grid in xscale
+    init_grid = LinRange(xscale[1](lims[1]), xscale[1](lims[2]), start_gridpoints)
+    # Current count of number mindpoints
+    num_gridpoints = start_gridpoints
+
+    # Current left point
+    xleft = init_grid[1]
+    xleft_identity = xscale[2](xleft)
+    leftvalue = func(xleft_identity)
+    # The full set of gridpoints
+    grid = [xleft_identity,]
+    # Tuple with list of all values (Faster than list of tuples + reshaping)
+    values = ntuple(i -> [leftvalue[i],], length(yscales))
+    
+    for xright in init_grid[2:end]
+        # Scale back to identity
+        xright_identity = xscale[2](xright)
+        rightvalue = func(xright_identity)
+        # Refine (if nessesary) section (xleft,xright)
+        num_gridpoints = refine_grid!(values, grid, func, xleft, xright, leftvalue, rightvalue, xscale, yscales, maxgridpoints, mineps, num_gridpoints; reltol_dd, abstol_dd)
+        # We are now done with [xleft,xright]
+        # Continue to next segment
+        xleft = xright
+        leftvalue = rightvalue
+    end
+    return values, grid
+end
+
+function push_multiple!(vecs::Tuple, vals::Tuple)
+    push!.(vecs, vals)
+end
+
+# Given range (xleft, xright) potentially add more gridpoints. xleft is assumed to be already added, xright will be added
+function refine_grid!(values, grid, func, xleft, xright, leftvalue, rightvalue, xscale, yscales, maxgridpoints, mineps, num_gridpoints; reltol_dd=0.05, abstol_dd=1e-2)
+    # In scaled grid
+    midpoint = (xleft+xright)/2
+    # Scaled back
+    midpoint_identity = xscale[2](midpoint)
+    midvalue = func(midpoint_identity)
+
+    #mineps in scaled version, abs to avoid assuming monotonly increasing scale
+    if (abs(xright - xleft) >= mineps) && (num_gridpoints < maxgridpoints) && !is_almost_linear(xleft, midpoint, xright, leftvalue, midvalue, rightvalue, xscale, yscales; reltol_dd, abstol_dd)
+        num_gridpoints = refine_grid!(values, grid, func, xleft,    midpoint, leftvalue, midvalue,   xscale, yscales, maxgridpoints, mineps, num_gridpoints; reltol_dd, abstol_dd)
+        num_gridpoints = refine_grid!(values, grid, func, midpoint, xright,   midvalue,  rightvalue, xscale, yscales, maxgridpoints, mineps, num_gridpoints; reltol_dd, abstol_dd)
+    else
+        # No more refinement needed, add computed points
+        push!(grid, midpoint_identity)
+        push!(grid, xscale[2](xright))
+        push_multiple!(values, midvalue)
+        push_multiple!(values, rightvalue)
+        num_gridpoints += 2
+    end
+    return num_gridpoints
+end
+
+# TODO We can scale when saving instead of recomputing scaling
+# Svectors should also be faster in general
+function is_almost_linear(xleft, midpoint, xright, leftvalue, midvalue, rightvalue, xscale, yscales; reltol_dd=0.05, abstol_dd=1e-2)
+    # We assume that x2-x1 \approx  x3-x1, so we need to check that y2-y1 approx y3-y2
```

Comment:
```suggestion
    # We assume that x2-x1 \approx  x3-x2, so we need to check that y2-y1 approx y3-y2
```

Assume this is what is meant?